### DISCUSSION TITLE
Performance Issues When Combining Google Search, Qdrant, and Neo4j in LangChain with Llama 3.2 3B

### DISCUSSION THREAD

**Discussion by MrBinit** [community/community#147411] [New to GitHub]
I’m building a chatbot using LangChain and Llama 3.2 3B, with retrieval from:

Google Search: For real-time web information. Qdrant (Vector Database): For semantic search. Neo4j (Knowledge Graph): For structured relationship-based queries. Problem:

When I use Google Search + Qdrant or Google Search + Neo4j, the results are good. When combining all three tools, the response quality deteriorates.

query: tell me about King mahendra

What I’ve Tried:

Verified tools individually and in pairs. Prompt engineering to structure outputs. Debugged for conflicts or redundancy in retrieved data. Question: What might cause poor performance when combining all three tools, and how can I optimize integration for better results?

Any advice or best practices for multi-tool integration in LangChain would be appreciated!

![image](https://github.com/user-attachments/assets/0f43ed07-2742-4bca-bbec-be34819ea737)
result when I use all three tools.

![image](https://github.com/user-attachments/assets/a04ad896-8e8e-46c2-81fa-960f80d96c5b)
result with qdrant and google search.

![image](https://github.com/user-attachments/assets/c584095a-a4db-44d8-b3f9-47b410fae585)
result with knowledge graph and google search

**Comment by tuves** [comment]
Hi @MrBinit,

Thanks for being a part of the GitHub Community, we're glad you're here!

If you're looking for help for this specific topic, you might want to try asking for help somewhere that focuses on this project, such as the [langchain's repository](https://github.com/langchain-ai/langchain). It's possible that another GitHub user might have run into this same issue and can help, but the GitHub Community Discussions focuses primarily on topics related to GitHub itself or collaboration on project development and ideas. We want to make sure you’re getting the best support you can, but this space may not be the right place for this particular topic.

Best of luck!
